### PR TITLE
Allow selection of drawing order in CircuitDrawer (solves #333)

### LINKED ISSUE #333: circuitdrawer reordering gates?
I am using the tikz output for a circuit, but the gates are not drawn in the order I am sending them. How can I force the gates to be in the specified order?

### PR DESCRIPTION
**Pull request by alexandrupaler**
Included the possibility to draw the gates in the order they were added to the circuit.

Closes #333

### CODE REVIEW

File: projectq/backends/_circuits/_drawer.py [hunk 1]
```diff
@@ -244,7 +246,14 @@ def _print_cmd(self, cmd):
...
 
-    def get_latex(self):
+        """
+        The commands are stored on a per qubit basis in a very strange manner
+        In order to draw according to the order how they were added here
+        I am storing the first qubit/line which the commands affect
+        """
+        self._paler_command_order.append(all_lines[0])
+
+    def get_latex(self, ordered = False, sequential_gates = False):
```
Comment (Takishima) [inline projectq/backends/_circuits/_drawer.py]: Please document the new argument in the function docstring

File: projectq/backends/_circuits/_to_latex.py [hunk 2]
```diff
@@ -17,7 +17,7 @@
                           Measure, SqrtSwap, Swap, X, Z)
 
 
-def to_latex(circuit):
+def to_latex(circuit, paler_order = None, paler_one_gate_at_a_time = False):
```
Comment (Takishima) [inline projectq/backends/_circuits/_to_latex.py]: Same thing about `paler` which is not clear to me...
And please document the new argument in the function docstring.

File: projectq/backends/_circuits/_to_latex.py [hunk 3]
```diff
@@ -234,7 +245,7 @@ def __init__(self, settings, num_lines):
         self.op_count = [0] * num_lines
         self.is_quantum = [settings['lines']['init_quantum']] * num_lines
 
-    def to_tikz(self, line, circuit, end=None):
+    def to_tikz(self, line, circuit, end=None, one_gate_at_a_time = False):
```
Comment (Takishima) [inline projectq/backends/_circuits/_to_latex.py]: Please document the new function argument inside the docstring.

File: projectq/backends/_circuits/_drawer.py [hunk 4]
```diff
@@ -244,7 +246,14 @@ def _print_cmd(self, cmd):
         for l in all_lines:
             self._qubit_lines[l].append(item)
 
-    def get_latex(self):
+        """
+        The commands are stored on a per qubit basis in a very strange manner
+        In order to draw according to the order how they were added here
+        I am storing the first qubit/line which the commands affect
+        """
+        self._paler_command_order.append(all_lines[0])
```
[the remark this answers is not in the document]
Comment (alexandrupaler) [reply projectq/backends/_circuits/_drawer.py]: new PR addresses this comment

File: projectq/backends/_circuits/_drawer.py (hunk 1 above)
Comment (alexandrupaler) [reply projectq/backends/_circuits/_drawer.py]: new PR addresses this comment

File: projectq/backends/_circuits/_to_latex.py (hunk 2 above)
Comment (alexandrupaler) [reply projectq/backends/_circuits/_to_latex.py]: new PR addresses this comment

File: projectq/backends/_circuits/_to_latex.py (hunk 3 above)
Comment (alexandrupaler) [reply projectq/backends/_circuits/_to_latex.py]: new PR addresses this comment

**Comment by alexandrupaler** [comment]
@Takishima  I am fine with signing a copyright form. Please send me the form.

For the tests, I had to change the test files a bit, because the lambdas did not have the necessary number of params

**Comment by Takishima** [comment]
> @Takishima I am fine with signing a copyright form. Please send me the form.
> 
> For the tests, I had to change the test files a bit, because the lambdas did not have the necessary number of params

To sign the CLA, you need to send an email to info@projectq.ch requesting it.

File: projectq/backends/_circuits/_to_latex_test.py [hunk 5]
```diff
@@ -47,7 +47,7 @@ def test_tolatex():
     old_footer = _to_latex._footer
 
     _to_latex._header = lambda x: "H"
-    _to_latex._body = lambda x, y: x
+    _to_latex._body = lambda x, settings, command_order, one_gate_at_a_time: x
```
Comment (Takishima) [inline projectq/backends/_circuits/_to_latex_test.py]: Rename it once you address the first comments.

File: projectq/backends/_circuits/_to_latex_test.py [hunk 6]
```diff
@@ -104,7 +104,7 @@ def test_large_gates():
     drawer = _drawer.CircuitDrawer()
     eng = MainEngine(drawer, [])
     old_tolatex = _drawer.to_latex
-    _drawer.to_latex = lambda x: x
+    _drawer.to_latex = lambda x, command_order, one_gate_at_a_time: x
```
Comment (Takishima) [inline projectq/backends/_circuits/_to_latex_test.py]: Rename it once you address the first comments.

File: projectq/backends/_circuits/_to_latex_test.py [hunk 7]
```diff
@@ -136,7 +136,7 @@ def test_body():
...
```
Comment (Takishima) [inline projectq/backends/_circuits/_to_latex_test.py]: Rename it once you address the first comments.

File: projectq/backends/_circuits/_to_latex_test.py [hunk 8]
```diff
@@ -186,7 +186,7 @@ def test_qubit_allocations_at_zero():
...
```
Comment (Takishima) [inline projectq/backends/_circuits/_to_latex_test.py]: Rename it once you address the first comments.

File: projectq/backends/_circuits/_to_latex_test.py [hunk 9]
```diff
@@ -219,7 +219,7 @@ def test_qubit_lines_classicalvsquantum1():
...
```
Comment (Takishima) [inline projectq/backends/_circuits/_to_latex_test.py]: Rename it once you address the first comments.

File: projectq/backends/_circuits/_drawer.py [hunk 10]
```diff
@@ -244,7 +252,15 @@ def _print_cmd(self, cmd):
         for l in all_lines:
             self._qubit_lines[l].append(item)
 
-    def get_latex(self):
+        """
+        The commands are stored on a per qubit basis in a very strange manner
+        In order to draw according to the order how they were added here
+        I am storing the first qubit/line which the commands affect
+        See also description in the __init__
+        """
```
Comment (Takishima) [inline projectq/backends/_circuits/_drawer.py]: This comment is not needed here.

File: projectq/backends/_circuits/_drawer_test.py [hunk 11]
```diff
@@ -32,7 +32,7 @@
 
 def test_drawer_getlatex():
     old_latex = _drawer.to_latex
-    _drawer.to_latex = lambda x: x
+    _drawer.to_latex = lambda x, command_order, one_gate_at_a_time: x
```
Comment (Takishima) [inline projectq/backends/_circuits/_drawer_test.py]: Rename it once you address the first comments.

File: projectq/backends/_circuits/_to_latex.py [hunk 12]
```diff
@@ -17,7 +17,7 @@
                           Measure, SqrtSwap, Swap, X, Z)
 
 
-def to_latex(circuit):
+def to_latex(circuit, command_order = None, one_gate_at_a_time = False):
```
Comment (Takishima) [inline projectq/backends/_circuits/_to_latex.py]: Rename it once you address the first comments.

**Comment by Takishima** [comment]
I will do a new review once the comments above have been addressed; overall it is starting to look better and we will be able to merge soon hopefully.

File: projectq/backends/_circuits/_to_latex.py [hunk 13]
```diff
@@ -182,22 +237,39 @@ def _header(settings):
...
 
     conv = _Circ2Tikz(settings, len(circuit))
-    for line in range(len(circuit)):
-        code.append(conv.to_tikz(line, circuit))
+
+    to_where = None
+    if drawing_order is None:
+        drawing_order = list(range(len(circuit)))
+    else:
+        to_where = 1
```
[the remark this answers is not in the document]
Comment (alexandrupaler) [reply projectq/backends/_circuits/_to_latex.py]: I will prepare a test and commit once I have it.

I have not received a response or any CLA form from info@projectq.ch

[the remark this answers is not in the document]
Comment (Takishima) [reply projectq/backends/_circuits/_to_latex.py]: > I will prepare a test and commit once I have it.

Awesome, thanks.


> I have not received a response or any CLA form from [info@projectq.ch](mailto:info@projectq.ch)

Don't worry, I can confirm that we received your email. We'll get you to sign the CLA form eventually.

[the remark this answers is not in the document]
Comment (alexandrupaler) [reply projectq/backends/_circuits/_to_latex.py]: Added tests and also found an issue which I solved.

**Comment by Takishima** [comment]
Closes #333

**Pull request merged**